.github/scripts/filter_test_configs.py

```diff
@@ -0,0 +1,146 @@
+
+CIFLOW_PREFIX = "ciflow/"
+
+# Same as shard names
+VALID_TEST_CONFIG_LABELS = {f"{CIFLOW_PREFIX}{label}" for label in {
+    "backwards_compat",
+    "crossref",
+    "default",
+    "deploy",
+    "distributed",
+    "docs_tests",
+    "dynamo",
+    "force_on_cpu",
+    "functorch",
+    "jit_legacy",
+    "multigpu",
+    "nogpu_AVX512",
+    "nogpu_NO_AVX2",
+    "slow",
+    "xla",
+}}
+
+def parse_args() -> Any:
+    from argparse import ArgumentParser
+    parser = ArgumentParser("Filter all test configurations and keep only requested ones")
+    parser.add_argument("--test-matrix", type=str, required=True, help="the original test matrix")
+    parser.add_argument("--pr-number", type=str, help="the pull request number")
+    parser.add_argument("--tag", type=str, help="the associated tag if it exists")
+    return parser.parse_args()
+
+
+def get_labels(pr_number: int) -> Set[str]:
+    """
+    Dynamical get the latest list of labels from the pull request
+    """
+    # From https://docs.github.com/en/actions/learn-github-actions/environment-variables
+    PYTORCH_REPO = os.environ.get("GITHUB_REPOSITORY", "pytorch/pytorch")
+    PYTORCH_GITHUB_API = f"https://api.github.com/repos/{PYTORCH_REPO}"
+    GITHUB_TOKEN = os.environ["GITHUB_TOKEN"]
+
+    REQUEST_HEADERS = {
+        "Accept": "application/vnd.github.v3+json",
+        "Authorization": "token " + GITHUB_TOKEN,
+    }
+
+    response = requests.get(
+        f"{PYTORCH_GITHUB_API}/issues/{pr_number}/labels",
+        headers=REQUEST_HEADERS,
+    )
+
+    if response.status_code != requests.codes.ok:
+        warnings.warn(f"Failed to get the labels for #{pr_number} (status code {response.status_code})")
+        return set()
+
+    return {label.get("name") for label in response.json() if label.get("name")}
+
+
+def filter(test_matrix: Dict[str, List[Any]], labels: Set[str]) -> Dict[str, List[Any]]:
+    """
+    Select the list of test config to run from the test matrix. The logic works
+    as follows:
+
+    If the PR has one or more labels as specified in the VALID_TEST_CONFIG_LABELS set, only
+    these test configs will be selected.  Note that regular ciflow labels like ciflow/trunk
+    don't count here, only test configs.
+
+    If the PR has none of the ciflow label, all tests are run as usual.
+    """
+
+    filtered_test_matrix: Dict[str, List[Any]] = {
+        "include": []
+    }
+
+    for entry in test_matrix.get("include", []):
+        config_name = entry.get("config", "")
+        if not config_name:
+            continue
+
+        label = f"{CIFLOW_PREFIX}{config_name.strip().lower()}"
+        if label in labels:
+            print(f"Select {config_name} because label {label} is presented in the pull request by the time the test starts")
+            filtered_test_matrix["include"].append(entry)
+
+    valid_test_config_labels = labels.intersection(VALID_TEST_CONFIG_LABELS)
+
+    if filtered_test_matrix["include"] or valid_test_config_labels:
```

Comment:
Why is this an `or`?

Corner cases this appears to allow: if a test has an invalid test config lables, and the PR gets both the valid and invalid label on it, then we'll run the test with the invalid labels

I'm guessing the behavior you want is: "If there are no tests with a valid label, but a the pr author requested only running tests with a specific valid label, then don't run any tests".

Reply:
Let me try to explain while an `OR` is used here because the second part `valid_test_config_labels` is used in a different way.  Let me know if that makes sense to you.

First of all, `filtered_test_matrix` already has all the valid configs to be run. I should be able to just return it right away unless it's empty (then all tests will be run).  That's the first part of the OR statement.

However, there is an edge case here where we do want to return an empty `filtered_test_matrix` instead. Take [linux-vulkan-bionic-py3_7-clang9-test](https://github.com/pytorch/pytorch/blob/master/.github/workflows/pull.yml#L147-L157) as an example, you can see that it has already `default` test config.  So when I set the `functorch` label, the filtered test matrix will be empty and I want it to stay that way instead of running all tests. This is the second part of the OR statement.

May be an easier way to understand this is to write the if statement in a different way:

```
if not filtered_test_matrix["include"] and not valid_test_config_labels:
  # filtered_test_matrix is empty AND there is no valid label, run everything
  return test_matrix
else:
  return filtered_test_matrix
```